libbeat/publisher/queue/diskqueue/docs/on-disk-structures.md

```diff
@@ -0,0 +1,76 @@
+# Disk Queue On Disk Structures
+
+The disk queue is a directory on disk that contains files.  Each
+file is called a segment.  The name of the file is the segment id in
+base 10 with the ".seg" suffix.  For example: "42.seg".  Each segment
+contains multiple frames.  Each frame contains one event.
+
+There are currently 3 versions of the disk queue, and the current code
+base is able to write versions 1 & 2, while it is able to read version
+0, 1, and 2.
+
+## Version 0
+
+In version 0, the segments are made up of a header, followed by
+frames.  The header contains one field which is an unsigned 32-bit
+integer in little-endian byte order, which signifies the version number.
+
+![Segment Schema Version 0](./schemaV0.svg)
+
+The frames for version 0, consist of a header, followed by the
+serialized event and a footer.  The header contains one field which is
+the size of the frame, which is an unsigned 32-bit integer in
+little-endian byte order.  The serialization format is JSON.  The
+footer contains 2 fields, the first of which is a checksum which is an
+unsigned 32-bit integer in little-endian format, followed by a repeat
+of the size from the header.
+
+![Frame Version 0](./frameV0.svg)
+
+## Version 1
+
+In version 1, the segments are made up of a header, followed by
+frames.  The header contains two fields.  The first field in the
+version number, which is an unsigned 32-bit integer in little-endian
+format.  The second field is a count of the number of frames in the
+segment, which is an unsigned 32-bit integer in little-endian format.
+
+![Segment Schema Version 1](./schemaV1.svg)
+
+The frames for version 1, consist of a header, followed by the
+serialized event and a footer.  The header contains one field which is
+the size of the frame, which is an unsigned 32-bit integer in
+little-endian format.  The serialization format is CBOR.  The footer
+contains 2 fields, the first of which is a checksum which is an
+unsigned 32-bit integer in little-endian format, followed by a repeat
+of the size from the header.
+
+![Frame Version 1](./frameV1.svg)
+
+## Version 2
```

Comment:
Just to clarify.  I think you are proposing a new schema (V2) that has a new header field for encryption, and we would just write V2 schema. Does that sound right?  I don't think we can keep it V1 since we need to have something that indicates encryption or not.

If so, I'm not adverse to that from a technical standpoint (my original plan was similar).  But ... the feedback from the design doc was not to go that way.

@cmacknz @jlind23 thoughts on changing design at this stage?  I'm OK either way, but I want to be sure that the code is something the team wants to support in the future.

Reply:
I think I agree with @faec in the sense that it would be nice to be able to just delete the V1 schema handling entirely in the future, rather than maintaining it indefinitely as the format used when a user doesn't want encryption. 

If we can unify both implementations that is easier from a future maintenance perspective, assuming that can be done without any major tradeoffs. I think @faec made a reasonable argument that the case where an attacker has gotten access to the files on disk with enough permissions to modify them. They can always just delete the segment files without bothering to modify them. As long as they can't silently modify the events themselves we are fine. 

Ideally want we want in the end is a disk queue that optionally supports compression and/or encryption to account for every users security + performance use cases.

Reply:
sounds good, I'll rework and do a new PR.  I'm going to leave this one so we can compare.

Reply:
new PR #31961 with proposed changes, so we can compare & contrast.